util/ranger/detacher.go

```diff
@@ -876,3 +877,386 @@ func MergeDNFItems4Col(ctx sessionctx.Context, dnfItems []expression.Expression)
+// (2) the first param of "IN" function shoulde be a column not a expression like `a + b`
+// (3) the rest params of "IN" function all shoulde be constant
+// (4) the first param of "IN" function shoulde be the column in the expression of first index field.
+//     e.g. uk(tidb_shard(a), a). If the condtions is `WHERE b in (1, 2, 3)`, the first param of "IN" function
+//     is `b` that's not the column in `tidb_shard(a)`.
+// @param  sf	"IN" function, e.g. `a IN (1, 2, 3)`
+func NeedAddColumn4InCond(cols []*expression.Column, accessCond []expression.Expression, sf *expression.ScalarFunction) bool {
+	if cols == nil || accessCond == nil || sf == nil {
+		return false
+	}
+
+	if accessCond[0] != nil {
+		return false
+	}
+
+	fields := GetFieldsFromExpr(cols[0].VirtualExpr.(*expression.ScalarFunction))
+
+	c, ok := sf.GetArgs()[0].(*expression.Column)
+	if !ok {
+		return false
+	}
+
+	for _, arg := range sf.GetArgs()[1:] {
+		if _, ok := arg.(*expression.Constant); !ok {
+			return false
+		}
+	}
+
+	if len(fields) != 1 ||
+		!fields[0].Equal(nil, c) {
+		return false
+	}
+
+	return true
+}
+
+// GetFieldsFromExpr get all fields from input expresion virtaulExpr
+func GetFieldsFromExpr(virtaulExpr *expression.ScalarFunction) []*expression.Column {
+	var fields []*expression.Column
+
+	if virtaulExpr == nil {
+		return fields
+	}
+
+	for _, arg := range virtaulExpr.GetArgs() {
+		if sf, ok := arg.(*expression.ScalarFunction); ok {
+			fields = append(fields, GetFieldsFromExpr(sf)...)
+		} else if c, ok := arg.(*expression.Column); ok {
+			if !c.InColumnArray(fields) {
+				fields = append(fields, c)
+			}
+		}
+	}
+
+	return fields
+}
+
+// IsValidShardIndex Check whether the definition of shard index is valid. The form of index
+// shoulde like `index(tidb_shard(a), a, ....)`.
+// 1) the column count shoudle be >= 2
+// 2) the first column shoudle be tidb_shard(xxx)
+// 3) the parameter of tidb_shard shoudle be a column that is the second column of index
+// @param[in] cols        the columns of shard index, such as [tidb_shard(a), a, ...]
+// @retval - if the shard index is valid return true, otherwise return false
+func IsValidShardIndex(cols []*expression.Column) bool {
+	// definition of index shoulde like the form: index(tidb_shard(a), a, ....)
+	if len(cols) < 2 {
+		return false
+	}
+
+	// the first coulmn of index must be GC column and the expr must be tidb_shard
+	if !expression.GcColumnExprIsTidbShard(cols[0].VirtualExpr) {
+		return false
+	}
+
+	shardFunc, _ := cols[0].VirtualExpr.(*expression.ScalarFunction)
+
+	argCount := len(shardFunc.GetArgs())
+	if argCount != 1 {
+		return false
+	}
+
+	// parameter of tidb_shard must be the second column of the input index columns
+	col, ok := shardFunc.GetArgs()[0].(*expression.Column)
+	if !ok || !col.Equal(nil, cols[1]) {
+		return false
+	}
+
+	return true
+}
+
+// IsEqOrInFunc check whetere it si equal or in function
+func IsEqOrInFunc(expr expression.Expression) bool {
```

Comment:
Is this function only used for testing? If so, we may move it into the test package.

Reply:
No, it is used in the normal code